Avoid unnecessary instantiations in DispatchToVectorized

benbarsdell opened the pull request:
- This avoids unnecessary template instantiations when vectorizing with 16-bit and larger data types, significantly reducing compile times for some kernels.
- It should also help with the Windows compilation issue discussed here:
  https://github.com/tensorflow/tensorflow/issues/54276#issuecomment-1032186006
- No functional change.

cc @nluehr

benbarsdell committed:
Avoid unnecessary instantiations in DispatchToVectorized

- This avoids unnecessary template instantiations when vectorizing with
  16-bit and larger data types, significantly reducing compile times for
  some kernels.
- It should also help with the Windows compilation issue discussed here:
  https://github.com/tensorflow/tensorflow/issues/54276#issuecomment-1032186006
- No functional change.

reedwm commented on tensorflow/core/util/gpu_kernel_helper.h:
```diff
@@ -314,6 +314,29 @@ int64_t MinAlignmentOf(Args... args) {
+  }
+};
+template <template <int vec_size> class Functor>
+struct DispatchToVectorizedHelper<0, Functor> {
```
I think it's clearer for the base case to be VecSize=1 instead of VecSize=0, since I think the VecSize=0 case will never actually be executed.

benbarsdell replied:
I've made this change. I got confused for a bit because it turns out there is a tested path that calls this with `max_vec_size=0`. That was happening because it was calling `MinAlignmentOf(nullptr, ...)` due to its input pointer being unused. I've fixed that too by treating zero/nullptr as infinitely aligned. I also added some error checks.

reedwm reviewed (commented):
@cheshire can you review? This looks right to me but I am not enough of a C++ wizard to know if this is the best way to do this

benbarsdell committed:
Fix small issues with DispatchToVectorized

- Changes the base case of template recursion from VecSize=0 to
  VecSize=1, because VecSize=0 will never be reached.
- Adds handling of `alignment_of(zero/nullptr)`, which should be treated
  as infinitely aligned.
- Adds checks for preconditions.

gbaned commented:
@reedwm Can you please review this PR ? Thank you!

gbaned commented:
@benbarsdell  Can you please address Ubuntu Sanity errors? Thank you!

benbarsdell commented:
Rebased, which seems to have fixed the CI (I think it was broken due to https://github.com/tensorflow/tensorflow/issues/55494, but I guess the CI doesn't automatically pick up that fix?).
